lib/composite_cache_store.rb

```diff
@@ -4,120 +4,132 @@
+    @layers = layers
   end
 
   def cleanup(...)
-    outer.cleanup(...)
-    inner.cleanup(...)
+    layers.each { |store| store.cleanup(...) }
   end
 
   def clear(...)
-    outer.clear(...)
-    inner.clear(...)
+    layers.each { |store| store.clear(...) }
   end
 
   def decrement(...)
-    outer.decrement(...)
-    inner.decrement(...)
+    layers.each { |store| store.decrement(...) }
   end
 
   def delete(...)
-    outer.delete(...)
-    inner.delete(...)
+    layers.each { |store| store.delete(...) }
   end
 
   def delete_matched(...)
-    outer.delete_matched(...)
-    inner.delete_matched(...)
+    layers.each { |store| store.delete_matched(...) }
   end
 
   def delete_multi(...)
-    outer.delete_multi(...)
-    inner.delete_multi(...)
+    layers.each { |store| store.delete_multi(...) }
   end
 
   def exist?(...)
-    outer.exist?(...) || inner.exist?(...)
+    layers.each do |store|
+      return true if store.exist?(...)
+    end
+    false
   end
 
   def fetch(*args, &block)
-    outer.fetch(*args) do
-      inner.fetch(*args, &block)
+    f = ->(store) do
+      return store.fetch(*args, &block) if store == layers.last
+      store.fetch(*args) { f.call(layers[layers.index(store) + 1]) }
     end
+    f.call(layers.first)
   end
 
   def fetch_multi(*args, &block)
-    outer.fetch_multi(*args) do
-      inner.fetch_multi(*args, &block)
+    fm = ->(store) do
+      return store.fetch_multi(*args, &block) if store == layers.last
+      store.fetch_multi(*args) { fm.call(layers[layers.index(store) + 1]) }
     end
+    fm.call(layers.first)
   end
 
-  # write
   def increment(...)
-    outer.increment(...)
-    inner.increment(...)
+    layers.each { |store| store.increment(...) }
   end
 
   def mute
-    outer.mute do
-      inner.mute do
-        yield
-      end
+    m = ->(store) do
+      return store.mute { yield } if store == layers.last
+      store.mute { m.call(layers[layers.index(store) + 1]) }
     end
+    m.call(layers.first)
   end
 
   def read(*args)
-    outer.fetch(*args) do
-      inner.read(*args)
+    r = ->(store) do
+      return store.read(*args) if store == layers.last
+      store.fetch(*args) { r.call(layers[layers.index(store) + 1]) }
     end
+    r.call(layers.first)
   end
 
   def read_multi(...)
-    result = outer.read_multi(...)
-    result = inner.read_multi(...) if result.blank?
-    result
+    layers.each do |store|
+      result = store.read_multi(...)
+      return result if result.present?
+    end
+    nil
   end
 
   def silence!
-    outer.silence!
-    inner.silence!
+    layers.each { |store| store.silence! }
   end
 
+  # Only applies expiration options to the outermost cache
+  # Inner caches use their global expiration options
   def write(name, value, options = nil)
     options ||= {}
-    outer.write(name, value, options.except(:expires_in)) # ? accept expires_in if less than outer.config[:expires_in] ?
-    inner.write(name, value, options)
+    layers.each do |store|
```

Comment:
not for this PR, but I wonder if we should, instead of doing this imperatively, add a `Refinement` of sorts to each cache layer, that will propagate `write` and `write_multi` to the layers below after a successful `write`/`write_multi` of its own.

Reply:
Interesting idea. I'm certainly open to exploring this option.